app/src/main/java/io/apicurio/registry/rest/v2/GroupsResourceImpl.java

```diff
@@ -645,10 +663,58 @@ private ArtifactMetaData createArtifactWithRefs(String groupId, ArtifactType xRe
 
         // TODO do something with the optional user-provided Version
 
+        if (fromURL != null) {
+            Client client = null;
+            try {
+                client = JAXRSClientUtil.getJAXRSClient(restConfig.getDownloadSkipSSLValidation());
+            } catch (KeyManagementException kme) {
+                throw new RuntimeException(kme);
+            } catch (NoSuchAlgorithmException nsae) {
+                throw new RuntimeException(nsae);
+            }
+
+            // 1. Registry issues HTTP HEAD request to the target URL.
+            List<Object> contentLengthHeaders = client
+                    .target(fromURL)
+                    .request()
+                    .head()
+                    .getHeaders()
+                    .get("Content-Length");
+
+            if (contentLengthHeaders == null || contentLengthHeaders.size() < 1) {
+                throw new BadRequestException("Requested resource URL does not provide 'Content-Length' in the headers");
+            }
+
+            // 2. According to HTTP specification, target server must return Content-Length header.
+            int contentLength = Integer.parseInt(contentLengthHeaders.get(0).toString());
+
+            // 3. Registry analyzes value of Content-Length to check if file with declared size could be processed securely.
+            if (contentLength > restConfig.getDownloadMaxSize()) {
+                throw new BadRequestException("Requested resource is bigger than " + restConfig.getDownloadMaxSize() + " and cannot be downloaded.");
+            }
+
+            // 4. Finally, registry issues HTTP GET to the target URL and fetches only amount of bytes specified by HTTP HEAD from step 1.
+            data = new BufferedInputStream(client
+                    .target(fromURL)
+                    .request()
+                    .get()
+                    .readEntity(InputStream.class), contentLength);
+        }
+
         ContentHandle content = ContentHandle.create(data);
         if (content.bytes().length == 0) {
             throw new BadRequestException(EMPTY_CONTENT_ERROR_MESSAGE);
         }
+
+        // Mitigation for MITM attacks, verify that the artifact is the expected one
+        if (xRegistryArtifactSHA != null) {
+            String calculatedSha = Hashing.sha256().hashString(content.content(), StandardCharsets.UTF_8).toString();
+
+            if (!calculatedSha.equals(xRegistryArtifactSHA)) {
+                throw new BadRequestException("Provided Artifact SHA doesn't match with the content");
+            }
+        }
```

Comment:
Why does this address the MITM attack?  If the attacker can change the payload, can't they also just change the SHA as well?

Reply:
A malicious actor can replace the server providing the file(especially if not using `https`), while the `sha` comes along with the request (and is not hosted on the remote hosting server).

Reply:
Ah I see - so useful in the `fromUrl` use-case more than in other cases.

Reply:
very correct 👍